reframe/core/decorators.py

```diff
@@ -27,7 +27,84 @@
 from reframe.utility.versioning import VersionValidator
 
 
-def _register_test(cls, args=None):
+class TestRegistry:
+    '''Regression test registry.
+
+    The tests are stored in a dictionary where the test class is the key
+    and the constructor arguments for the different instantiations of the
+    test are stored as the dictionary value as a list of (args, kwargs)
+    tuples.
+
+    For backward compatibility reasons, the registry also contains a set of
+    tests to be skipped. The machinery related to this should be dropped with
+    the ``required_version`` decorator.
+    '''
+
+    def __init__(self):
+        self._tests = dict()
+        self._skip_tests = set()
+
+    @classmethod
+    def create(cls, test, *args, **kwargs):
+        obj = cls()
+        obj.add(test, *args, **kwargs)
+        return obj
+
+    def add(self, test, *args, **kwargs):
+        if test in self._tests:
+            self._tests[test].append((args, kwargs))
+        else:
+            self._tests[test] = [(args, kwargs)]
```

Comment:
We could write this in a simpler way:

```suggestion
        self._tests.setdefault(test, [])
        self._tests[test].append((args, kwargs))
```